python/ray/tune/api_v2/tuner.py

```diff
@@ -0,0 +1,231 @@
+                elif isinstance(v, int):
+                    # CV settings
+                    pass
+                elif isinstance(v, list):
+                    if not all([isinstance(v_item, int) for v_item in v]) and not all(
+                        [isinstance(v_item, Dataset) for v_item in v]
+                    ):
+                        raise TuneError(
+                            "Wrongly formatted dataset param passed in Tune!"
+                        )
+                    if len(v) > 0 and isinstance(v[0], Dataset):
+                        dataset_dict[k] = [v_item.fully_executed() for v_item in v]
+                else:
+                    # We shouldn't be expecting anything here.
+                    raise TuneError("Unexpected dataset param passed in.")
+
+        if DATASETS in self.param_space:
+            ds = self.param_space[DATASETS]
+            if isinstance(ds, Dataset):
+                self.param_space[DATASETS] = ds.fully_executed()
+            elif isinstance(ds, dict):
+                _helper(ds)
+            else:
+                # We shouldn't be expecting anything here.
+                raise TuneError("Unexpected dataset param passed in.")
+
+    @staticmethod
+    def _convert_trainable(trainable: Any):
+        if isinstance(trainable, ConvertibleToTrainable):
+            trainable = trainable.as_trainable()
+        else:
+            trainable = trainable
+        return trainable
+
+    def _get_experiment_checkpoint_dir(self):
+        # TODO: This needs to be generalized.
+        path = "/Users/xwjiang/ray_results/tuner_resume"
+        os.makedirs(path)
+        return path
+
+    def fit(self):
+        trainable = self._convert_trainable(self.trainable)
+        if not self.is_restored:
+            param_space = copy.deepcopy(self.param_space)
+            analysis = self._fit_internal(trainable, param_space)
+        else:
+            analysis = self._fit_resume(trainable)
+
+        return analysis
+
+    def _fit_internal(self, trainable, param_space):
+        """Fitting for a fresh Tuner.
+
+        `Tuner.experiment_path` is assigned from `analysis.experiment_dir`."""
+        analysis = run(
+            trainable,
+            config={"run_config": self.run_config, **param_space},
+            name=self.name,
+            callbacks=self.callbacks,
+        )
+        return analysis
+
+    def _fit_resume(self, trainable):
+        """Fitting for a restored Tuner.
+
+        `Tuner.experiment_path` is already set in this case."""
+        assert self.experiment_path
+        # E.g. /home/ray/ray_results/experiment_name
+        experiment_dir = os.path.dirname(self.experiment_path + os.sep)
+        # E.g. experiment_name
+        experiment_name = os.path.basename(experiment_dir)
+        # E.g. /home/ray/ray_results
+        local_dir = os.path.dirname(experiment_dir)
+
+        analysis = run(
+            trainable, name=experiment_name, local_dir=local_dir, resume=True
```

Comment:
The `callbacks` have to be passed again. We will probably have to make a decision here if users have to pass them again or not. Callbacks can have state, so we may have to include them in the experiment checkpoint.
Note that this is currently also not the case in `tune.run` so we don't necessarily have to do it right now.

Reply:
Two things I'm wondering:

1) Should we make "included in checkpoint" the line of what to include in the constructor args vs. the runtime args (i.e. `Tuner(arg=v)` vs `tune.run(arg=val)`? I.e. if callbacks are not checkpointed and thus don't retain their state, make them part of runtime arguments so users don't expect them to retain it on resume?

2) Should we think about a general purpose API for saving/loading checkpoints of arbitrary components, e.g. callbacks? So that we could write checkpointable callbacks without having to include them explicitly in tuner state